Services/Certificate/classes/API/Repository/class.ilUserDataRepository.php

```diff
@@ -0,0 +1,197 @@
+<?php declare(strict_types=1);
+/* Copyright (c) 1998-2019 ILIAS open source, Extended GPL, see docs/LICENSE */
+namespace Certificate\API\Repository;
+
+use Certificate\API\Data\ilUserCertificateData;
+use Certificate\API\Filter\UserDataFilter;
+
+/**
+ * @author  Niels Theen <ntheen@databay.de>
+ * @ilCtrl_Calls: ilUserDataRepository: ilUserCertificateApiGUI
+ */
+class ilUserDataRepository
+{
+    /**
+     * @var \ilDBInterface
+     */
+    private $database;
+
+    /**
+     * @var \ilLogger
+     */
+    private $logger;
+
+    /**
+     * @var null|string
+     */
+    private $defaultTitle;
+
+    /**
+     * @var \ilCtrl
+     */
+    private $controller;
+
+    /**
+     * @param \ilDBInterface $database
+     * @param \ilLogger      $logger
+     * @param \ilCtrl        $controller
+     * @param string|null    $defaultTitle
+     */
+    public function __construct(
+        \ilDBInterface $database,
+        \ilLogger $logger,
+        \ilCtrl $controller,
+        string $defaultTitle = null
+    ) {
+        $this->database = $database;
+        $this->logger = $logger;
+        $this->controller = $controller;
+
+        if (null === $defaultTitle) {
+            global $DIC;
+            $defaultTitle = $DIC->language()->txt('certificate_no_object_title');
+        }
+        $this->defaultTitle = $defaultTitle;
+    }
+
+    /**
+     * @param UserDataFilter $filter
+     * @param array          $ilCtrlStack
+     * @return array
+     */
+    public function getUserData(UserDataFilter $filter, array $ilCtrlStack) : array
```

Comment:
I will check this. I think an `ORDER BY` will be possible, but I am sure `Pagination` cannot be applied here, because the query results in multiple rows if an ILIAS object with a certificate has multiple references.